core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/util/QueryEvaluationUtil.java

```diff
@@ -152,8 +152,8 @@ public static boolean compareLiterals(Literal leftLit, Literal rightLit, Compare
 		IRI leftDatatype = leftLit.getDatatype();
 		IRI rightDatatype = rightLit.getDatatype();
 
-		XSD.Datatype leftXsdDatatype = leftLit.getXsdDatatype().orElse(null);
-		XSD.Datatype rightXsdDatatype = rightLit.getXsdDatatype().orElse(null);
+		XSD.Datatype leftXsdDatatype = Literals.getXsdDatatype(leftLit).orElse(null);
+		XSD.Datatype rightXsdDatatype = Literals.getXsdDatatype(leftLit).orElse(null);
```

Comment:
This is probably what is causing the test failures. Notice both methods call on leftLit